File: javascript/leetcode/153. Find Minimum in Rotated Sorted Array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMin = function (nums) {
  if (!nums.length) return -1;

  let start = 0;
  let end = nums.length - 1;

  while (start < end) {
    if (nums[start] < nums[end]) break;

    let mid = start + parseInt((end - start) / 2);

    if (nums[start] <= nums[mid]) start = ++mid;
    else end = mid;
  }

  return nums[start];
};

File: c++/leetcode/153. Find Minimum in Rotated Sorted Array.cpp
class Solution
{
public:
  int findMin(vector<int> &nums)
  {
    if (!nums.size())
      return -1;

    int start = 0;
    int end = nums.size() - 1;

    while (start < end)
    {
      if (nums[start] < nums[end])
        return nums[start];

      auto mid = start + ((end - start) / 2);

      if (nums[start] <= nums[mid])
      {
        start = ++mid;
      }
      else
      {
        end = mid;
      }
    }

    return nums[start];
  }
};